demos/projects/ST/stm32h745i-disco/cm7/port/azure_iot_flash_platform.c

```diff
@@ -0,0 +1,229 @@
+        return eAzureIoTErrorFailed;
+    }
+
+    mbedtls_md_context_t ctx;
+    mbedtls_md_type_t md_type = MBEDTLS_MD_SHA256;
+
+    mbedtls_md_init( &ctx );
+    mbedtls_md_setup( &ctx, mbedtls_md_info_from_type( md_type ), 0 );
+    mbedtls_md_starts( &ctx );
+
+    AZLogInfo( ( "Starting the mbedtls calculation: image size %d\r\n", pxAduImage->ulImageFileSize ) );
+
+    for( size_t ulOffset = 0; ulOffset < pxAduImage->ulImageFileSize; ulOffset += sizeof( ucPartitionReadBuffer ) )
+    {
+        ulReadSize = pxAduImage->ulImageFileSize - ulOffset < sizeof( ucPartitionReadBuffer ) ? pxAduImage->ulImageFileSize - ulOffset : sizeof( ucPartitionReadBuffer );
+
+        memcpy( ucPartitionReadBuffer, ( pxAduImage->xUpdatePartition + ulOffset ), ulReadSize );
+
+        mbedtls_md_update( &ctx, ( const unsigned char * ) ucPartitionReadBuffer, ulReadSize );
+    }
+
+    AZLogInfo( ( "mbedtls calculation completed\r\n" ) );
+
+    mbedtls_md_finish( &ctx, ucCalculatedHash );
+    mbedtls_md_free( &ctx );
+
+    if( memcmp( ucDecodedManifestHash, ucCalculatedHash, azureiotflashSHA_256_SIZE ) == 0 )
+    {
+        AZLogInfo( ( "SHAs match\r\n" ) );
+        xResult = eAzureIoTSuccess;
+    }
+    else
+    {
+        AZLogError( ( "SHAs do not match\r\n" ) );
+        AZLogInfo( ( "Wanted: " ) );
+
+        for( int i = 0; i < azureiotflashSHA_256_SIZE; ++i )
+        {
+            AZLogInfo( ( "%x", ucDecodedManifestHash[ i ] ) );
+        }
+
+        AZLogInfo( ( "\r\n" ) );
+        AZLogInfo( ( "Calculated: " ) );
+
+        for( int i = 0; i < azureiotflashSHA_256_SIZE; ++i )
+        {
+            AZLogInfo( ( "%x", ucCalculatedHash[ i ] ) );
+        }
+
+        AZLogInfo( ( "\r\n" ) );
+
+        xResult = eAzureIoTErrorFailed;
+    }
+
+    return xResult;
+}
+
+AzureIoTResult_t AzureIoTPlatform_EnableImage( AzureADUImage_t * const pxAduImage )
+{
+    FLASH_OBProgramInitTypeDef xOptionBytes;
+    AzureIoTResult_t xResult = eAzureIoTSuccess;
+
+    HAL_FLASH_Unlock();
+    HAL_FLASH_OB_Unlock();
+    HAL_FLASHEx_OBGetConfig( &xOptionBytes ); /* Get current optionbytes configuration */
+    xOptionBytes.OptionType = OPTIONBYTE_USER;
+    xOptionBytes.USERType = OB_USER_SWAP_BANK;
+
+    /* If swap is disabled, we are in bank 1 */
+    xOptionBytes.USERConfig = ( xOptionBytes.USERConfig & OB_SWAP_BANK_ENABLE ) == OB_SWAP_BANK_DISABLE
+                              ? OB_SWAP_BANK_ENABLE
+                              : OB_SWAP_BANK_DISABLE;
+
+    if( HAL_FLASHEx_OBProgram( &xOptionBytes ) != HAL_OK )
+    {
+        AZLogError( ( "Error setting option bytes\r\n" ) );
+        xResult = eAzureIoTErrorFailed;
+    }
+
+    HAL_FLASH_Lock();
+    HAL_FLASH_OB_Lock();
+
+    return xResult;
+}
+
+AzureIoTResult_t AzureIoTPlatform_ResetDevice( AzureADUImage_t * const pxAduImage )
+{
+    HAL_FLASH_Unlock();
+    HAL_FLASH_OB_Unlock();
+
+    HAL_FLASH_OB_Launch();
+
+    NVIC_SystemReset();
```

Comment:
On the H7, do you need this call? For the L4, OB_Launch resets the device. If it doesn't reset the device, it might be good to move OB_Launch to the enable image function

Reply:
Yes oddly enough it does. OB launch doesn't cause it to reboot even though I have set the option bytes as the sample has.